### PR TITLE
HIVE-22622: Hive allows to create a struct with duplicate attribute names

### PR DESCRIPTION
**Pull request by kasakrisz**
### What changes were proposed in this pull request?
Add a check for duplicated struct attribute identifiers and throw SemanticException with customized error message when found.

### Why are the changes needed?
Creating a table with a struct type column with duplicate attribute identifier and inserting records is allowed but later when querying from the table we cannot distinguish between the attributes of the struct has the same identifier.
In some cases (depending on table serde format) the query may fails. See jira for details.

### Does this PR introduce _any_ user-facing change?
Introduce new error code and message. Example:
```
FAILED: SemanticException [Error 10423]: Attribute "id" specified more than once in structured type.
```

### How was this patch tested?
1. Create new negative test:
```
mvn test -Dtest.output.overwrite -DskipSparkTests -Dtest=TestNegativeCliDriver -Dqfile=struct_attribute_uniqueness.q,struct_attribute_uniqueness_2.q -pl itests/qtest -Pitests
```

2. Reproduce query failure
```
CREATE TABLE person
(
    `id`      int,
    `address` struct<number:int,street:string,number:int>
)
    ROW FORMAT SERDE
        'org.apache.hadoop.hive.ql.io.orc.OrcSerde'
    STORED AS INPUTFORMAT
        'org.apache.hadoop.hive.ql.io.orc.OrcInputFormat'
        OUTPUTFORMAT
            'org.apache.hadoop.hive.ql.io.orc.OrcOutputFormat';

INSERT INTO person
VALUES (1, named_struct('number', 61, 'street', 'Terrasse', 'number', 62));
INSERT INTO person
VALUES (2, named_struct('number', 51, 'street', 'Terrasse', 'number', 52));

SELECT address.number FROM person;
```

### CODE REVIEW

**Review by zabetak — COMMENTED** [review]
Thanks for taking care of this @kasakrisz :)

If I remember well the underlying format (ORC, JSON, etc) of the table was throwing a different error when I was running the reproducer. 

Can you add a test for every supported format to ensure that we have a consistent behavior during CREATE?

I didn't check if it is in general supported in Hive but does the check capture duplicate attributes in nested struct types? Can you please add a test if that's the case.

File: common/src/java/org/apache/hadoop/hive/ql/ErrorMsg.java [hunk 1]
```diff
@@ -471,6 +471,7 @@
           "Not an ordered-set aggregate function: {0}. WITHIN GROUP clause is not allowed.", true),
   WITHIN_GROUP_PARAMETER_MISMATCH(10422,
           "The number of hypothetical direct arguments ({0}) must match the number of ordering columns ({1})", true),
+  AMBIGUOUS_STRUCT_FIELD(10423, "Struct field is not unique: {0}", true),
```
Comment (zabetak) [inline common/src/java/org/apache/hadoop/hive/ql/ErrorMsg.java:474]: nit: Usually we use "field" for row types and "attribute" for struct types. Plus it reads more natural if we inline the attribute name in the sentence: 

`Attribute \"{0}\" specified more than once in structured type.`

Comment (kasakrisz) [reply common/src/java/org/apache/hadoop/hive/ql/ErrorMsg.java]: Renamed `field` to `attribute`

**Comment by kasakrisz** [comment]
Adding a test case for various underlying formats doesn't make sense this case because the duplicate check is performed in the semantical analysis phase of the `create table` statement and the error message would be the same. So if a duplicate found the code flow doesn't reach the point where the table is actually created.

Added a test case for nested struct.
